### PR TITLE
Isa uses binding manager `add_constraint`.

### PR DESCRIPTION
**Pull request by dhatch**
PR checklist:

- [ ] Added changelog entry.

### CODE REVIEW

File: polar-core/src/bindings.rs [hunk 1]
```diff
@@ -311,3 +318,19 @@ impl BindingManager {
...
     // bindings
 }
+
+/// Get variable out of a term for ``add_constraint`` to determine where the
+/// constraint is stored.
+///
+/// If the term is a variable, the variable is returned.
+/// If the term is a dot expression, the VAR from VAR.field is returned.
+/// Otherwise, the term is returned.
+fn extract_variable(value: &Value) -> &Value {
```
Comment (samscott89) [inline polar-core/src/bindings.rs:328]: When does this happen? Are there cases when we call `add_constraint` on a term like:
`x.foo <= 1` ? Why wouldn't that be rewritten

File: polar-core/src/vm.rs [hunk 2]
```diff
@@ -947,6 +953,9 @@ impl PolarVirtualMachine {
             | (Value::RestVariable(l), Value::RestVariable(r)) => {
                 // Two variables.
                 match (self.variable_state(l), self.variable_state(r)) {
+                    // TODO (dhatch): This special case does not seem right to me.
```
Comment (samscott89) [inline polar-core/src/vm.rs:956]: Agreed.

File: polar-core/src/vm.rs [hunk 3]
```diff
@@ -961,41 +970,7 @@ impl PolarVirtualMachine {
...
-                        self.constrain(&e)?;
-                    }
-                    (VariableState::Partial(mut e), VariableState::Partial(f)) => {
-                        e.merge_constraints(f);
-                        let e = e.clone_with_new_constraint(
-                            op!(Isa, left.clone(), right.clone()).into_term(),
-                        );
-                        self.constrain(&e)?;
-                    }
+                    (_, _) => self.add_constraint(&term!(op!(Isa, left.clone(), right.clone())))?,
```
Comment (samscott89) [inline polar-core/src/vm.rs:973]: Nice

File: polar-core/src/vm.rs [hunk 4]
```diff
@@ -1007,10 +982,10 @@ impl PolarVirtualMachine {
                     left: x,
                     right: right.clone(),
                 })?,
-                VariableState::Cycle(c) => self.isa_expr(&cycle_constraints(c), left, right)?,
-                VariableState::Partial(e) => self.isa_expr(&e, left, right)?,
+                _ => self.isa_expr(left, right)?,
             },
             (_, Value::Variable(r)) | (_, Value::RestVariable(r)) => match self.variable_state(r) {
+                // TODO (dhatch): Is this case tested?
```
Comment (samscott89) [inline polar-core/src/vm.rs:988]: Hmmm, yes, I think so. Look at `test_non_instance_specializers`

I think the "legitimate" use case for this is something like:

```
allow(user: User { id: id, is_staff: True }, ... ) if ...
```

I.e you want the isa to bind variables in the pattern.

File: polar-core/src/vm.rs [hunk 5]
```diff
@@ -1096,48 +1070,43 @@ impl PolarVirtualMachine {
...
-                        }
-                        _ => unify,
-                    }
+                    let unify = op!(Unify, left, value);
+                    term!(unify)
                 };
 
-                let mut operation = operation.clone();
-
                 // Add all but the last field constraint to the operation.
```
Comment (samscott89) [inline polar-core/src/vm.rs:1087]: Comment is no longer correct.

How come the so-called "bind dance"  isn't required any more?

File: polar-core/src/vm.rs [hunk 6]
```diff
@@ -1096,48 +1070,43 @@ impl PolarVirtualMachine {
...
                 let var = left.value().as_symbol()?;
-                let simplified = simplify_partial(var, operation.clone().into_term());
+                // Get the existing partial on the LHS variable.
+                let partial = match self.variable_state(var) {
+                    VariableState::Partial(expr) => expr,
+                    VariableState::Cycle(c) => cycle_constraints(c),
+                    _ => panic!("Invariant violated. left must be a variable"),
+                };
+
+                let simplified = simplify_partial(var, partial.into_term());
```
Comment (samscott89) [inline polar-core/src/vm.rs:1105]: I'm not sure if this is tested or correct.

File: polar-core/src/vm.rs [hunk 7]
```diff
@@ -130,6 +130,11 @@ pub enum Goal {
     BindBatch {
         bindings: Rc<RefCell<BindingStack>>,
     },
+
+    /// Add a new constraint
+    AddConstraint {
```
Comment (samscott89) [inline polar-core/src/vm.rs:135]: Should this be called `Constrain` to match the convention of the other goals?

File: polar-core/src/vm.rs [hunk 8]
```diff
@@ -1171,50 +1138,29 @@ impl PolarVirtualMachine {
...
-                                VariableState::Partial(e) => {
-                                    unify = e.clone_with_new_constraint(unify.into_term());
-                                }
-                                VariableState::Unbound => (),
-                            }
-                        }
-                        _ => (),
-                    }
-                    operation = operation.clone_with_new_constraint(term!(unify));
+                    op!(Unify, left, v)
```
Comment (samscott89) [inline polar-core/src/vm.rs:1145]: Should this be `Isa`?

File: polar-core/src/vm.rs [hunk 9]
```diff
@@ -1517,15 +1463,16 @@ impl PolarVirtualMachine {
...
-                            self.push_goal(Goal::Unify { left, right })?;
-                        }
                         VariableState::Bound(value) => {
                             return Err(self.type_error(&left, format!("Can only assign to unbound variables, {} is bound to value {}.", var.to_polar(), value.to_polar())));
                         }
+                        // TODO should this one be allowed?
                         VariableState::Partial(e) => {
                             return Err(self.type_error(&left, format!("Can only assign to unbound variables, {} is bound to expression {}.", var.to_polar(), e.to_polar())));
                         }
+                        _ => {
```
Comment (samscott89) [inline polar-core/src/vm.rs:1473]: Nit: Personally I think it's clearer to have the extra variant case spelled out.

**Review by samscott89 — APPROVED** [review]
Nice, this looks like a good improvement.

File: polar-core/src/bindings.rs (hunk 1 above)
Comment (dhatch) [reply polar-core/src/bindings.rs:328]: Removed in new code

File: polar-core/src/vm.rs (hunk 2 above)
Comment (dhatch) [reply polar-core/src/vm.rs:956]: Alex removed in https://github.com/osohq/oso/pull/723/commits/0231cc74c6eaad57b362b40baffbd8026a2153c0

File: polar-core/src/vm.rs (hunk 6 above)
Comment (dhatch) [reply polar-core/src/vm.rs:1105]: There are some integration tests, but it's not well tested.

File: polar-core/src/vm.rs (hunk 5 above)
Comment (dhatch) [reply polar-core/src/vm.rs:1087]: Not sure, I have no idea what that comment was referring to.

File: polar-core/src/vm.rs (hunk 8 above)
Comment (dhatch) [reply polar-core/src/vm.rs:1145]: Yeah, it should

Comment (dhatch) [reply polar-core/src/vm.rs:1145]: Looks like the non constraint code uses Isa, not Unify.

**Pull request closed without merging**